dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/shared/GridAdaptor.java

```diff
@@ -83,10 +97,72 @@ public Grid createGrid( final List<GridHeader> headers, final Map<Column, List<O
                 rowsAdded = true;
             }
 
-            // Note that the header column must match the result map key.
-            grid.addHeader( header ).addColumn( columnRows );
+            if ( !commonQueryRequest.isSkipHeaders() )
+            {
+                // Note that the header column must match the result map key.
+                grid.addHeader( header );
+            }
+
+            grid.addColumn( columnRows );
+        }
+
+        if ( !commonQueryRequest.isSkipMeta() )
+        {
+            grid.setMetaData( getMetadata( teiQueryParams, commonQueryRequest ) );
         }
 
         return grid;
     }
+
+    /**
+     * returns metadata based on teiQueryParams and rendered by
+     * commonQueryRequest
+     *
+     * @param teiQueryParams
+     * @param commonQueryRequest
+     * @return
+     */
+    private static Map<String, Object> getMetadata( final TeiQueryParams teiQueryParams,
```

Comment:
Would be nice to have a unit test in GridAdaptorTest to test this method's logic.